test/integration/transaction-relay.ts

```diff
@@ -0,0 +1,88 @@
+import hre from "hardhat";
+import chai from "chai";
+import { solidity } from "ethereum-waffle";
+import cap from "chai-as-promised";
+import type { TransactionRelay } from "../../typechain/TransactionRelay";
+import { getDeployer, getStakingRewardsDistribution, getUser1, getUser2 } from "../../utils/DeployedContractsHelpers";
+import { StakingRewardsDistribution } from "../../typechain/StakingRewardsDistribution";
+import { SignerWithAddress } from "hardhat-deploy-ethers/dist/src/signers";
+
+chai.use(cap);
+
+chai.use(solidity);
+const { expect } = chai;
+
+async function GetTransactionRelay() {
+  const timelock = (await hre.ethers.getContract("TransactionRelay")) as TransactionRelay;
+  return timelock;
+}
+
+async function CreateTransactionAs(creator: SignerWithAddress) {
+  const transactionRelay = (await GetTransactionRelay()).connect(creator);
+
+  const stakingRewardsDistribution = await getStakingRewardsDistribution(hre);
+
+  const tx = await transactionRelay.createTransaction(
+    stakingRewardsDistribution.address,
+    0,
+    "",
+    stakingRewardsDistribution.interface.encodeFunctionData("setVestingRewardRatio", [13])
+  );
+
+  const receipt = await tx.wait();
+  const createTransactionEventsArgs = receipt.events?.find(e => e.event === "CreateTransaction")?.args;
+
+  let txHash: string;
+  if (createTransactionEventsArgs) {
+    txHash = createTransactionEventsArgs[0];
+  } else {
+    throw new Error("Missing event");
+  }
+
+  return txHash;
+}
+
+describe.only("Execute with transaction relay", () => {
+  let relay: TransactionRelay;
+  let stakingRewardsDistribution: StakingRewardsDistribution;
+  let deployer: SignerWithAddress;
+  let multisigImposter: SignerWithAddress;
+  let randomUser: SignerWithAddress;
+
+  beforeEach(async () => {
+    await hre.deployments.fixture();
```

Comment:
This setup exaplains the whole workflow and idea behind it:

1. TransactionRelay contract (probably not the best name) should become the owner of all of the other contracts
2. Multisig becomes TransactionRelay's ower
3. Anybody can store a transaction TransactionRelay (which is identified by the hash)
4. The hash is returned to the creator
5. We would have to change our tasks to call transaction relay 
6. The tasks would output transactions' ids (hashes) to the terminal
7. System admin callas TransactionRelay.executeTransaction() via Gnosis multi-sig (admin only needs a single ABI with a single function and a list of transaction hashes)

Possible improvements:

1. Merge transaction relay with timelock
2. Allow to execute multiple transactions an once
3. Create a dynamic framework which would allow to use transaction relay client object to work with any interface

Reply:
regarding 1 - This deployed TransactionRelay contract should be one of the signers of our multisigs, right?

Reply:
@AndrzejGisWideLab  About 2 - if the TransactionRelay contract will be the owner of all the other contracts - it means that we won't be able to perform any action via the gnosis UI. we will have to do all the interactions via the tasks and the TransactionRelay.
Am I right?

Reply:
@tkeidar yes. every action would have to be initiated with code. Possible solution might be to implement fallback function like proxy does:
https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/proxy/Proxy.sol 

This _should_ work